File: lib/data/services/local_storage_service.dart
import 'package:hive_flutter/hive_flutter.dart';
import 'package:shared_preferences/shared_preferences.dart';
import '../../core/constants/app_constants.dart';
import '../models/project_model.dart';
import '../models/export_settings_model.dart';

class LocalStorageService {
  static LocalStorageService? _instance;
  static LocalStorageService get instance =>
      _instance ?? (_instance = LocalStorageService._());
  LocalStorageService._();

  late SharedPreferences _prefs;
  late Box _settingsBox;
  late Box _projectsBox;
  late Box _favoritesBox;
  late Box _recentFilesBox;

  Future<void> init() async {
    await Hive.initFlutter();
    _prefs = await SharedPreferences.getInstance();
    _settingsBox = await Hive.openBox(AppConstants.settingsBox);
    _projectsBox = await Hive.openBox(AppConstants.projectsBox);
    _favoritesBox = await Hive.openBox(AppConstants.favoritesBox);
    _recentFilesBox = await Hive.openBox(AppConstants.recentFilesBox);
  }

  // Theme
  bool get isDarkMode => _prefs.getBool(AppConstants.kThemeMode) ?? true;
  Future<void> setDarkMode(bool value) =>
      _prefs.setBool(AppConstants.kThemeMode, value);

  // Onboarding
  bool get isOnboardingDone => _prefs.getBool(AppConstants.kOnboardingDone) ?? false;
  Future<void> setOnboardingDone() =>
      _prefs.setBool(AppConstants.kOnboardingDone, true);

  // Premium
  bool get isPremium => _prefs.getBool(AppConstants.kPremiumStatus) ?? false;
  Future<void> setPremium(bool value) =>
      _prefs.setBool(AppConstants.kPremiumStatus, value);

  // User ID cache
  String? get cachedUserId => _prefs.getString(AppConstants.kUserId);
  Future<void> setCachedUserId(String id) =>
      _prefs.setString(AppConstants.kUserId, id);
  Future<void> clearUserId() => _prefs.remove(AppConstants.kUserId);

  // Export settings
  ExportSettingsModel getExportSettings() {
    final map = _settingsBox.get('export_settings');
    if (map == null) return const ExportSettingsModel();
    return ExportSettingsModel.fromMap(Map<String, dynamic>.from(map as Map));
  }

  Future<void> saveExportSettings(ExportSettingsModel settings) =>
      _settingsBox.put('export_settings', settings.toMap());

  // Projects (local cache)
  List<ProjectModel> getCachedProjects() {
    final data = _projectsBox.values.toList();
    return data
        .map((e) => ProjectModel.fromMap(Map<String, dynamic>.from(e as Map)))
        .toList()
      ..sort((a, b) => (b.updatedAt ?? DateTime(0))
          .compareTo(a.updatedAt ?? DateTime(0)));
  }

  Future<void> saveProject(ProjectModel project) =>
      _projectsBox.put(project.id, project.toMap());

  Future<void> deleteProject(String id) => _projectsBox.delete(id);

  ProjectModel? getProject(String id) {
    final data = _projectsBox.get(id);
    if (data == null) return null;
    return ProjectModel.fromMap(Map<String, dynamic>.from(data as Map));
  }

  // Favorites
  Set<String> getFavoriteIds() =>
      Set<String>.from(_favoritesBox.values.map((e) => e.toString()));

  Future<void> addFavorite(String projectId) =>
      _favoritesBox.put(projectId, projectId);

  Future<void> removeFavorite(String projectId) =>
      _favoritesBox.delete(projectId);

  bool isFavorite(String projectId) =>
      _favoritesBox.containsKey(projectId);

  // Recent files
  List<String> getRecentFiles() =>
      _recentFilesBox.values.map((e) => e.toString()).toList();

  Future<void> addRecentFile(String filePath) async {
    final files = getRecentFiles();
    files.remove(filePath);
    files.insert(0, filePath);
    await _recentFilesBox.clear();
    for (final f in files.take(20)) {
      await _recentFilesBox.add(f);
    }
  }

  // Auto-save setting
  bool get isAutoSaveEnabled => _prefs.getBool(AppConstants.kAutoSave) ?? true;
  Future<void> setAutoSave(bool value) =>
      _prefs.setBool(AppConstants.kAutoSave, value);

  // Clear all data
  Future<void> clearAll() async {
    await _prefs.clear();
    await _settingsBox.clear();
    await _projectsBox.clear();
    await _favoritesBox.clear();
    await _recentFilesBox.clear();
  }

  // User-specific clear (keep settings)
  Future<void> clearUserData() async {
    await _projectsBox.clear();
    await _favoritesBox.clear();
    await _recentFilesBox.clear();
    await clearUserId();
  }

  Future<void> close() async {
    await _settingsBox.close();
    await _projectsBox.close();
    await _favoritesBox.close();
    await _recentFilesBox.close();
  }
}

File: services/storage/localStorage.service.ts
import { MMKV } from 'react-native-mmkv';
import { Project, ExportSettings, defaultExportSettings } from '../../types';
import { AppConstants } from '../../constants/AppConstants';

const storage = new MMKV({ id: 'erick-storage' });

class LocalStorageService {
  // ─── Theme ─────────────────────────────────────────────────────────────────
  get isDarkMode(): boolean {
    return storage.getBoolean(AppConstants.THEME_MODE) ?? true;
  }
  setDarkMode(value: boolean) {
    storage.set(AppConstants.THEME_MODE, value);
  }

  // ─── Onboarding ────────────────────────────────────────────────────────────
  get isOnboardingDone(): boolean {
    return storage.getBoolean(AppConstants.ONBOARDING_DONE) ?? false;
  }
  setOnboardingDone() {
    storage.set(AppConstants.ONBOARDING_DONE, true);
  }

  // ─── Premium ───────────────────────────────────────────────────────────────
  get isPremium(): boolean {
    return storage.getBoolean(AppConstants.PREMIUM_STATUS) ?? false;
  }
  setPremium(value: boolean) {
    storage.set(AppConstants.PREMIUM_STATUS, value);
  }

  // ─── User ID ───────────────────────────────────────────────────────────────
  get userId(): string | null {
    return storage.getString(AppConstants.USER_ID) ?? null;
  }
  setUserId(id: string) {
    storage.set(AppConstants.USER_ID, id);
  }
  clearUserId() {
    storage.delete(AppConstants.USER_ID);
  }

  // ─── Export Settings ───────────────────────────────────────────────────────
  getExportSettings(): ExportSettings {
    const raw = storage.getString('export_settings');
    if (!raw) return defaultExportSettings;
    try {
      return { ...defaultExportSettings, ...JSON.parse(raw) };
    } catch {
      return defaultExportSettings;
    }
  }
  saveExportSettings(settings: ExportSettings) {
    storage.set('export_settings', JSON.stringify(settings));
  }

  // ─── Projects (local cache) ────────────────────────────────────────────────
  getCachedProjects(): Project[] {
    const raw = storage.getString('cached_projects');
    if (!raw) return [];
    try {
      const list: Project[] = JSON.parse(raw);
      return list.sort(
        (a, b) => new Date(b.updatedAt).getTime() - new Date(a.updatedAt).getTime()
      );
    } catch {
      return [];
    }
  }
  saveProject(project: Project) {
    const projects = this.getCachedProjects();
    const idx = projects.findIndex((p) => p.id === project.id);
    if (idx >= 0) projects[idx] = project;
    else projects.unshift(project);
    storage.set('cached_projects', JSON.stringify(projects));
  }
  deleteProject(id: string) {
    const projects = this.getCachedProjects().filter((p) => p.id !== id);
    storage.set('cached_projects', JSON.stringify(projects));
  }
  getProject(id: string): Project | null {
    return this.getCachedProjects().find((p) => p.id === id) ?? null;
  }

  // ─── Favorites ─────────────────────────────────────────────────────────────
  getFavoriteIds(): Set<string> {
    const raw = storage.getString('favorites');
    if (!raw) return new Set();
    try {
      return new Set(JSON.parse(raw) as string[]);
    } catch {
      return new Set();
    }
  }
  addFavorite(id: string) {
    const favs = this.getFavoriteIds();
    favs.add(id);
    storage.set('favorites', JSON.stringify([...favs]));
  }
  removeFavorite(id: string) {
    const favs = this.getFavoriteIds();
    favs.delete(id);
    storage.set('favorites', JSON.stringify([...favs]));
  }
  isFavorite(id: string): boolean {
    return this.getFavoriteIds().has(id);
  }

  // ─── Recent files ──────────────────────────────────────────────────────────
  getRecentFiles(): string[] {
    const raw = storage.getString('recent_files');
    if (!raw) return [];
    try {
      return JSON.parse(raw) as string[];
    } catch {
      return [];
    }
  }
  addRecentFile(uri: string) {
    const files = this.getRecentFiles().filter((f) => f !== uri);
    files.unshift(uri);
    storage.set('recent_files', JSON.stringify(files.slice(0, 20)));
  }

  // ─── Auto-save ─────────────────────────────────────────────────────────────
  get isAutoSave(): boolean {
    return storage.getBoolean(AppConstants.AUTO_SAVE) ?? true;
  }
  setAutoSave(value: boolean) {
    storage.set(AppConstants.AUTO_SAVE, value);
  }

  // ─── Clear ─────────────────────────────────────────────────────────────────
  clearUserData() {
    storage.delete('cached_projects');
    storage.delete('favorites');
    storage.delete('recent_files');
    this.clearUserId();
    this.setPremium(false);
  }

  clearAll() {
    storage.clearAll();
  }
}

export const localStorage = new LocalStorageService();
